src/librustc/ty/layout.rs

```diff
@@ -648,6 +648,25 @@ impl Scalar {
             false
         }
     }
+
+    /// Returns a range suitable to be passed to LLVM for range metadata.
+    pub fn range_metadata<C: HasDataLayout>(&self, cx: C) -> Option<Range<u128>> {
```

Comment:
First off, please do not put LLVM special-cases outside of `librustc_trans`. Secondly, could the check be performed by having a way to get the full range for the size of the primitive in question?

We should *really* assert, somewhere, that `start & mask == start` and `end & mask == end`, because niche computation relies on it - #49981 already fixed an instance of this invariant not being upheld.

Because then you can just go `prim.value.full_valid_range(cx) == prim.valid_range` and that's your check.

Reply:
> First off, please do not put LLVM special-cases outside of librustc_trans.

Where should I put that in `librustc_trans`, and how? A free-standing function? A method on some `C` defined in `librustc_trans`? A trait implemented for `Scalar`?

> Secondly, could the check be performed by having a way to get the full range for the size of the primitive in question?

Nope, `0..=255` isn't equal to `1..=0`, so I doubt this is a good idea.

Reply:
Alright, could this function be named `valid_range_exclusive` instead, and have the `Option` replaced with a `r.start == r.end` in `rustc_trans`?

Reply:
Done!